### PR TITLE
[CodeCamp2023-527] Add pixel contrast cross entropy loss

### PR DESCRIPTION
**Pull request by BLUE-coconut**
Thanks for your contribution and we appreciate it a lot. The following instructions would make your pull request more healthy and more easily get feedback. If you do not understand some items, don't worry, just make the pull request and seek help from maintainers.

## Motivation

Please describe the motivation of this PR and the goal you want to achieve through this PR.

## Modification

Please briefly describe what modification is made in this PR.

## BC-breaking (Optional)

Does the modification introduce changes that break the backward-compatibility of the downstream repos?
If so, please describe how it breaks the compatibility and how the downstream projects should modify their code to keep compatibility with this PR.

## Use cases (Optional)

If this PR introduces a new feature, it is better to list some use cases here, and update the documentation.

## Checklist

1. Pre-commit or other linting tools are used to fix the potential lint issues.
2. The modification is covered by complete unit tests. If not, please add more unit test to ensure the correctness.
3. If the modification has potential influence on downstream projects, this PR should be tested with downstream projects, like MMDet or MMDet3D.
4. The documentation has been modified accordingly, like docstring or example tutorials.

### CODE REVIEW

File: mmseg/models/decode_heads/hrnetconstrast_head.py [hunk 1]
```diff
@@ -0,0 +1,174 @@
...
+            in_channels=self.in_channels, out_channels=proj_n, proj=proj_mode)
+
+    def forward(self, inputs):
+        inputs = self._transform_inputs(inputs)
+        output = {}
+        output['seg'] = self.seghead(inputs)
+
+        output['proj'] = self.projhead(inputs)
+
+        return output
```
Comment (xiexinch) [inline mmseg/models/decode_heads/hrnetconstrast_head.py:107]: The output should be a list.

File: mmseg/models/decode_heads/hrnetconstrast_head.py [hunk 2]
```diff
@@ -0,0 +1,174 @@
...
+
+    def loss_by_feat(self, seg_logits: dict,
+                     batch_data_samples: SampleList) -> dict:
```
Comment (xiexinch) [inline mmseg/models/decode_heads/hrnetconstrast_head.py:110]: The seg_logits should be a list.

File: mmseg/models/decode_heads/hrnetconstrast_head.py [hunk 3]
```diff
@@ -0,0 +1,174 @@
...
+                    seg_logits,
+                    seg_label,
+                    weight=seg_weight,
+                    ignore_index=self.ignore_index)
+            else:
+                loss[loss_decode.loss_name] += loss_decode(
+                    seg_logits,
+                    seg_label,
+                    weight=seg_weight,
+                    ignore_index=self.ignore_index)
```
Comment (xiexinch) [inline mmseg/models/decode_heads/hrnetconstrast_head.py:148]: Following the implementation of PaddleSeg, the input of different loss methods should be different.

https://github.com/PaddlePaddle/PaddleSeg/blob/2c8c35a8949fef74599f5ec557d340a14415f20d/paddleseg/core/train.py#L39

**Review by xiexinch — COMMENTED** [review]
Hi @BLUE-coconut,
Thanks for your contribution! Could you move your code to the `project` folder, we'll really appreciate it!

File: mmseg/models/losses/pixel_contrast_cross_entropy_loss.py [hunk 4]
```diff
@@ -0,0 +1,287 @@
+# Copyright (c) OpenMMLab. All rights reserved.
+import torch
+import torch.nn as nn
+
+from mmseg.registry import MODELS
+
+
+def hard_anchor_sampling(X, y_hat, y, ignore_index, max_views, max_samples):
```
Comment (xiexinch) [inline mmseg/models/losses/pixel_contrast_cross_entropy_loss.py:8]: We should add type hints.

File: mmseg/models/losses/pixel_contrast_cross_entropy_loss.py [hunk 5]
```diff
@@ -0,0 +1,287 @@
...
+            Sampled samplers of a class. Default: 100.
+    """
+
+    def __init__(self,
+                 loss_name='loss_pixel_contrast_cross_entropy',
+                 temperature=0.1,
+                 base_temperature=0.07,
+                 ignore_index=255,
+                 max_samples=1024,
+                 max_views=100):
```
Comment (xiexinch) [inline mmseg/models/losses/pixel_contrast_cross_entropy_loss.py:217]: We should add type hints.

File: projects/pixel_contrast_cross_entropy_loss/README.md [hunk 6]
```diff
@@ -0,0 +1,37 @@
+# Pixel contrast cross entropy loss
+
+[Exploring Cross-Image Pixel Contrast for Semantic Segmentation](https://arxiv.org/pdf/2101.11939.pdf)
```
Comment (xiexinch) [inline projects/pixel_contrast_cross_entropy_loss/README.md:3]:
```suggestion
> [Exploring Cross-Image Pixel Contrast for Semantic Segmentation](https://arxiv.org/pdf/2101.11939.pdf)
```

**Review by xiexinch — APPROVED** [review]

File: projects/pixel_contrast_cross_entropy_loss/hrnetconstrast_head.py [hunk 7]
```diff
@@ -0,0 +1,183 @@
...
+                For details on the values of these keys see
+                `mmseg/datasets/pipelines/formatting.py:PackSegInputs`.
+            test_cfg (dict): The testing config.
+
+        Returns:
+            Tensor: Outputs segmentation logits map.
+        """
+        seg_logits = self.forward(inputs)
+
+        return self.predict_by_feat(seg_logits, batch_img_metas)
```
Comment (xiexinch) [inline projects/pixel_contrast_cross_entropy_loss/hrnetconstrast_head.py:183]: Might overwrite the `predict_by_feat` method

File: projects/pixel_contrast_cross_entropy_loss/hrnetconstrast_head.py [hunk 8]
```diff
@@ -0,0 +1,183 @@
...
+        if proj_n <= 0:
+            raise KeyError('proj_n must >0')
+        if drop_p < 0 or drop_p > 1 or not isinstance(drop_p, float):
+            raise KeyError('drop_p must be a float >=0')
+        self.proj_n = proj_n
+
+        self.seghead = SegmentationHead(
+            in_channels=self.in_channels,
+            out_channels=self.num_classes,
+            drop_prob=drop_p)
```
Comment (xiexinch) [inline projects/pixel_contrast_cross_entropy_loss/hrnetconstrast_head.py:99]: Might use `MODEL.build()` and the `SegmentationHead` head is not necessary.

File: projects/pixel_contrast_cross_entropy_loss/hrnetconstrast_head.py [hunk 9]
```diff
@@ -0,0 +1,183 @@
...
+            nn.Dropout2d(drop_prob),
+            nn.Conv2d(in_channels, out_channels, kernel_size=1),
+        )
+
+    def forward(self, x: Tensor) -> Tensor:
+        return self.seg(x)
+
+
+@MODELS.register_module()
+class ContrastHead(BaseDecodeHead):
```
Comment (xiexinch) [inline projects/pixel_contrast_cross_entropy_loss/hrnetconstrast_head.py:76]: The `conv_seg` module in `BaseDecodeHead` is not used and it might cause a distributed training error. It suggests to inherit from `nn.Module`.

File: projects/pixel_contrast_cross_entropy_loss/hrnetconstrast_head.py [hunk 10]
```diff
@@ -11,6 +11,7 @@
 from mmcv.cnn import ConvModule
 from torch import Tensor
 
+from mmseg.models.builder import HEADS, build_head
```
Comment (xiexinch) [inline projects/pixel_contrast_cross_entropy_loss/hrnetconstrast_head.py:14]: Might remove this line.

File: projects/pixel_contrast_cross_entropy_loss/hrnetconstrast_head.py [hunk 11]
```diff
@@ -0,0 +1,284 @@
...
+                 in_channels=[18, 36, 72, 144],
+                 in_index=(0, 1, 2, 3),
+                 channels=sum([18, 36, 72, 144]),
+                 input_transform='resize_concat',
+                 proj_n=256,
+                 proj_mode='convmlp',
+                 drop_p=0.1,
+                 dropout_ratio=-1,
+                 num_classes=19,
+                 norm_cfg=norm_cfg,
```
Comment (xiexinch) [inline projects/pixel_contrast_cross_entropy_loss/hrnetconstrast_head.py:100]: Can't find the definition of norm_cfg.

Comment (xiexinch) [inline projects/pixel_contrast_cross_entropy_loss/hrnetconstrast_head.py:100]:
```suggestion
                 norm_cfg=dict(type='BN'),
```

File: projects/pixel_contrast_cross_entropy_loss/hrnetconstrast_head.py [hunk 12]
```diff
@@ -0,0 +1,284 @@
...
+                     num_convs=2,
+                     concat_input=True,
+                     dropout_ratio=0.1,
+                     num_classes=150,
+                     norm_cfg=norm_cfg,
+                     align_corners=False,
+                     loss_decode=dict(
+                         type='CrossEntropyLoss',
+                         use_sigmoid=False,
+                         loss_weight=1.0))**kwargs):
```
Comment (xiexinch) [inline projects/pixel_contrast_cross_entropy_loss/hrnetconstrast_head.py:127]: The decode head config might move to config file.

File: projects/pixel_contrast_cross_entropy_loss/hrnetconstrast_head.py (hunk 11 above)
Comment (xiexinch) [reply projects/pixel_contrast_cross_entropy_loss/hrnetconstrast_head.py:100]: This parameter seems not in use.

File: projects/pixel_contrast_cross_entropy_loss/hrnetconstrast_head.py [hunk 13]
```diff
@@ -0,0 +1,284 @@
...
+                     dict(
+                         type='PixelContrastCrossEntropyLoss',
+                         base_temperature=0.07,
+                         temperature=0.1,
+                         ignore_index=255,
+                         max_samples=1024,
+                         max_views=100,
+                         loss_weight=0.1),
+                     dict(type='CrossEntropyLoss', loss_weight=1.0)
+                 ],
```
Comment (xiexinch) [inline projects/pixel_contrast_cross_entropy_loss/hrnetconstrast_head.py:112]: Same to the decode head, it might move to the config file.

**Review by xiexinch — CHANGES_REQUESTED** [review]

**Review by xiexinch — APPROVED** [review]
